cmd/sourced/cmd/web.go

```diff
@@ -32,54 +33,177 @@ func init() {
+// 2vudhlzztdlbg_bblfsh-web_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /tini -- bblfshd                 Up                      0.0.0.0:9432->9432/tcp
+// 2vudhlzztdlbg_bblfsh_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /bin/sh -c sleep 10s && gh ...   Up
+// 2vudhlzztdlbg_ghsync_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   ./init.sh                        Up                      0.0.0.0:3306->3306/tcp
+// 2vudhlzztdlbg_gitbase_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /bin/dumb-init -- /bin/sh  ...   Up
+// 2vudhlzztdlbg_gitcollector_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /home/entrypoint.sh start- ...   Up                      0.0.0.0:10000->10000/tcp, 0.0.0.0:4040->4040/tcp, 8888/tcp
+// 2vudhlzztdlbg_gsc_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5433->5432/tcp
+// 2vudhlzztdlbg_metadatadb_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5432->5432/tcp
+// 2vudhlzztdlbg_postgres_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh redis ...   Up                      0.0.0.0:6379->6379/tcp
+// 2vudhlzztdlbg_redis_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /entrypoint.sh                   Up (health: starting)   0.0.0.0:8088->8088/tcp
+// 2vudhlzztdlbg_sourced-ui_1
+func runMonitor(ch chan<- error) {
+	runMonitorService := func(service string, ch chan<- error) {
+		for {
+			var stdout bytes.Buffer
+			if err := compose.RunWithIO(context.Background(),
+				os.Stdin, &stdout, nil, "ps", service); err != nil {
+				ch <- errors.Wrapf(err, "cannot get status service %s", service)
+				return
+			}
+
+			matches := stateExtractor.FindAllStringSubmatch(
+				normalizeNewLine(strings.TrimSpace(stdout.String())), -1)
+			for _, match := range matches {
+				state := match[1]
+
+				if strings.HasPrefix(state, "Exit") {
+					if service != "ghsync" && service != "gitcollector" {
+						ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+						return
+					}
+
+					returnCode := state[len("Exit "):len(state)]
+					if returnCode != "0" {
+						ch <- fmt.Errorf("service '%s' exited with return code: %s",
+							service, returnCode)
+						return
+					}
+
+					continue
+				}
+
+				if state != "Up" {
+					ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+					return
+				}
+
+			}
+
+			time.Sleep(1 * time.Second)
+		}
 	}
 
-	return nil
+	go func() {
```

Comment:
let's move `go` out of this function, it will be easier to understand and would allow to use the function in sync mode if needed

Reply:
it's mentioned [here](https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions) but not very clear. There was a better explanation somewhere (most probably in go blog or somewhere else, I can't find it right now). The idea is it's better to keep functions blocking and if you need to run them in non-blocking way the caller can do it which is much easier to reason about.

Reply:
Seems fair, thanks for link!

Reply:
Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/789f447283e70ac23ec272e92d8c5064dfa635f6).